Log improvements

sledgehammer999 opened the pull request:
1. Now the UI log can filter events
![log-menu](https://cloud.githubusercontent.com/assets/273315/12539687/18493c2a-c301-11e5-9c9e-68b77444e040.jpg)

2. Now there is the possibility to save the logs to disk
![log-conf](https://cloud.githubusercontent.com/assets/273315/12539691/2ba1b00e-c301-11e5-8a22-81b360d103c0.jpg)

Sample file output:
```
(N) 2016-01-25T01:00:25 - qBittorrent v3.4.0alpha started
(I) 2016-01-25T01:00:25 - qBittorrent is trying to listen on any interface port: 7777
(N) 2016-01-25T01:00:25 - Peer ID: -qB3400-
(N) 2016-01-25T01:00:25 - HTTP User-Agent is 'qBittorrent v3.4.0alpha'
(I) 2016-01-25T01:00:25 - Anonymous mode [OFF]
(I) 2016-01-25T01:00:25 - DHT support [ON]
(I) 2016-01-25T01:00:25 - PeX support [ON]
(I) 2016-01-25T01:00:25 - Local Peer Discovery support [ON]
(I) 2016-01-25T01:00:25 - Encryption support [ON]
(N) 2016-01-25T01:00:25 - Embedded Tracker [OFF]
(I) 2016-01-25T01:00:25 - UPnP / NAT-PMP support [ON]
(I) 2016-01-25T01:00:25 - GeoIP database loaded. Type: GeoLite2-Country. Build time: Τρι Ιαν 5 21:35:34 2016.
(N) 2016-01-25T01:00:25 - The Web UI is listening on port 30077
(N) 2016-01-25T01:00:25 - Options were saved successfully.
(I) 2016-01-25T01:00:27 - qBittorrent is successfully listening on interface 0.0.0.0 port: TCP/7777
(I) 2016-01-25T01:00:27 - qBittorrent is successfully listening on interface 0.0.0.0 port: TCP_SSL/4433
(I) 2016-01-25T01:00:27 - qBittorrent is successfully listening on interface :: port: TCP/7777
(I) 2016-01-25T01:00:27 - qBittorrent is successfully listening on interface :: port: TCP_SSL/4433
(I) 2016-01-25T01:00:27 - qBittorrent is successfully listening on interface 0.0.0.0 port: UDP/7777
(I) 2016-01-25T01:00:27 - qBittorrent is successfully listening on interface 0.0.0.0 port: TCP/7777
(I) 2016-01-25T01:00:27 - qBittorrent is successfully listening on interface :: port: TCP/7777
(I) 2016-01-25T01:00:27 - qBittorrent is successfully listening on interface 0.0.0.0 port: UDP/7777
(W) 2016-01-25T01:00:29 - Couldn't download favicon for URL 'http://nyaatorrents.info/favicon.ico'. Reason: The remote host name was not found (invalid hostname)
```

sledgehammer999 commented:
@qbittorrent/qbittorrent-frequent-contributors

glassez commented on src/base/filelogger.cpp:
```diff
@@ -0,0 +1,108 @@
+    m_manage = pref->fileLogManaged();
+    m_size = pref->fileLogMaxSize();
+    m_fileLogAction = static_cast<FileLogAction>(pref->fileLogAction());
+
```
Extra blank line.

glassez commented on src/base/filelogger.cpp:
```diff
@@ -0,0 +1,108 @@
+#include "logger.h"
+#include "utils/fs.h"
+
+FileLogger::FileLogger()
```
This subsystem (file logger) is not part of Base. It's part of App which uses Base service Logger (like LogListWidget is part of GUI and it also uses Logger).

sledgehammer999 replied:
Ok, I'll move it. I generally had in my mind base == non-gui and had forgotten about the other places.

glassez commented on src/base/preferences.cpp:
```diff
@@ -1054,6 +1056,115 @@ void Preferences::setExecutionLogEnabled(bool b)
     setValue("Preferences/ExecutionLog/enabled", b);
 }
 
+bool Preferences::executionLogNormalType() const
```
Isn't that a little wasteful? Maybe it's better to save one value and interpret it as a set of flags?

sledgehammer999 replied:
Sure. Can you recommend a type? I am thinking `char` and setting its first 4bits to on/off. (needs extra code though)

glassez replied:
Maybe you should try QFlags?

sledgehammer999 replied:
Good idea but QVariant doesn't have a constructor for that. I am thinking of a QBitArray.

glassez replied:
QVariant has generic fromValue(). QFlags is better since it allows to use something like Log::Normal | Log::Warning.

glassez commented on src/gui/executionlog.cpp:
```diff
@@ -35,18 +35,22 @@
     , ui(new Ui::ExecutionLog)
-    , m_msgList(new LogListWidget(MAX_LOG_MESSAGES))
-    , m_peerList(new LogListWidget(MAX_LOG_MESSAGES))
+    , m_peerList(new LogListWidget(MAX_LOG_MESSAGES, true, true, true, true))
```
What about default values for constructor params?

glassez replied:
By and large, this class should not contain any logic associated with handling log messages (and only the initialization/configuration of interface elements). This should be moved to the appropriate widget code.

glassez replied:
To call widget classes in the form of  ```<Something>Widget``` (eg.  ExecutionLogWidget) is a good thing.

glassez replied:
>By and large, this class should not contain any logic associated with handling log messages (and only the initialization/configuration of interface elements). This should be moved to the appropriate widget code

Forget it if you want. Actually, I understand why you're using this proxy code. You just want to have a simple widget (LogListWidget) that would operate only with some abstract string messages and their categories. I just somewhat confused by the fact that categories aren't used to working with peers... Nothing. Just do the default values for constructor parameters.

sledgehammer999 replied:
Exactly. The only "optimization" I could do is move the 2 formatting functions from `ExecutionLog` into the `LogListWidget` class as slots. Then in `ExecutionLog` connect the correct instance to the correct singal/slot.
So it is a tradeoff: A simple `ExecutionLog` or a simple `LogListWidget`. I don't really care which.

glassez commented on src/gui/mainwindow.ui:
```diff
@@ -95,12 +95,23 @@
     <property name="title">
      <string>&amp;View</string>
     </property>
+    <widget class="QMenu" name="menu_Log">
```
Please use the same rule for names as in c++ files. ```menuLog```, ```actionExecutionLogs```, etc.
Existing names in MainWindow I fixed in one of my last PRs.

sledgehammer999 replied:
Sure. That name was auto generated by QtDesigner...

Chocobo1 commented:
![pic](https://cloud.githubusercontent.com/assets/9395168/12581374/230db638-c471-11e5-9b6e-456b932101f2.jpg)
I wonder if the items in the red box can be put on the same line, i.e. don't use groupbox here

glassez commented:
>I wonder if the items in the red box can be put on the same line

I think Yes. Like on the line above.

sledgehammer999 commented:
>I wonder if the items in the red box can be put on the same line

How to signify enable/disable? My thought is to convert the "After" QLabel to a QCheckBox.

Chocobo1 commented:
>How to signify enable/disable? My thought is to convert the "After" QLabel to a QCheckBox.

I just think putting them on the same line is enough.

Chocobo1 commented on src/gui/options.ui:
```diff
@@ -481,6 +481,113 @@
+              </layout>
+             </widget>
+            </item>
+            <item>
```
```diff
diff --git a/src/gui/options.ui b/src/gui/options.ui
index 86fe030..4a14ec3 100644
--- a/src/gui/options.ui
+++ b/src/gui/options.ui
@@ -161,9 +161,9 @@
            <property name="geometry">
             <rect>
              <x>0</x>
-             <y>-191</y>
-             <width>454</width>
-             <height>733</height>
+             <y>-237</y>
+             <width>480</width>
+             <height>775</height>
             </rect>
            </property>
            <layout class="QVBoxLayout" name="verticalLayout_9">
@@ -488,6 +488,9 @@
               <property name="checkable">
                <bool>true</bool>
               </property>
+              <property name="checked">
+               <bool>true</bool>
+              </property>
               <layout class="QVBoxLayout" name="verticalLayout_11">
                <item>
                 <layout class="QHBoxLayout" name="horizontalLayout_7">
@@ -503,12 +506,6 @@
                  </item>
                  <item>
                   <widget class="QPushButton" name="browseFileLogDir">
-                   <property name="sizePolicy">
-                    <sizepolicy hsizetype="Fixed" vsizetype="Fixed">
-                     <horstretch>0</horstretch>
-                     <verstretch>0</verstretch>
-                    </sizepolicy>
-                   </property>
                    <property name="maximumSize">
                     <size>
                      <width>25</width>
@@ -516,73 +513,75 @@
                     </size>
                    </property>
                    <property name="text">
-                    <string>...</string>
+                    <string notr="true">...</string>
                    </property>
                   </widget>
                  </item>
                 </layout>
                </item>
                <item>
-                <widget class="QGroupBox" name="checkFileLogManage">
-                 <property name="title">
-                  <string>Manage File Size</string>
+                <layout class="QHBoxLayout" name="horizontalLayout_8">
+                 <property name="topMargin">
+                  <number>0</number>
                  </property>
-                 <property name="checkable">
-                  <bool>true</bool>
-                 </property>
-                 <layout class="QHBoxLayout" name="horizontalLayout">
-                  <item>
-                   <widget class="QLabel" name="label_11">
+                 <item>
+                  <widget class="QCheckBox" name="checkFileLogManage">
+                   <property name="text">
+                    <string>Manage file size:</string>
+                   </property>
+                  </widget>
+                 </item>
+                 <item>
+                  <widget class="QLabel" name="label_11">
+                   <property name="text">
+                    <string comment="After 10 MB backup the log file">After</string>
+                   </property>
+                  </widget>
+                 </item>
+                 <item>
+                  <widget class="QSpinBox" name="spinFileLog">
+                   <property name="suffix">
+                    <string> MB</string>
+                   </property>
+                   <property name="minimum">
+                    <number>1</number>
+                   </property>
+                   <property name="maximum">
+                    <number>1000</number>
+                   </property>
+                   <property name="value">
+                    <number>10</number>
+                   </property>
+                  </widget>
+                 </item>
+                 <item>
+                  <widget class="QComboBox" name="comboFileLog">
+                   <item>
                     <property name="text">
-                     <string comment="After 10 MB backup the log file">After</string>
-                    </property>
-                   </widget>
-                  </item>
-                  <item>
-                   <widget class="QSpinBox" name="spinFileLog">
-                    <property name="suffix">
-                     <string> MB</string>
+                     <string comment="After 10 MB Backup the log file">Backup</string>
                     </property>
-                    <property name="minimum">
-                     <number>1</number>
-                    </property>
-                    <property name="maximum">
-                     <number>1000</number>
-                    </property>
-                    <property name="value">
-                     <number>10</number>
-                    </property>
-                   </widget>
-                  </item>
-                  <item>
-                   <widget class="QComboBox" name="comboFileLog">
-                    <property name="sizePolicy">
-                     <sizepolicy hsizetype="MinimumExpanding" vsizetype="Fixed">
-                      <horstretch>0</horstretch>
-                      <verstretch>0</verstretch>
-                     </sizepolicy>
-                    </property>
-                    <item>
-                     <property name="text">
-                      <string comment="After 10 MB Backup the log file">Backup</string>
-                     </property>
-                    </item>
-                    <item>
-                     <property name="text">
-                      <string comment="After 10 MB Delete the log file">Delete</string>
-                     </property>
-                    </item>
-                   </widget>
-                  </item>
-                  <item>
-                   <widget class="QLabel" name="label_12">
+                   </item>
+                   <item>
                     <property name="text">
-                     <string comment="After 10 MB backup the log file">the log file</string>
+                     <string comment="After 10 MB Delete the log file">Delete</string>
                     </property>
-                   </widget>
-                  </item>
-                 </layout>
-                </widget>
+                   </item>
+                  </widget>
+                 </item>
+                 <item>
+                  <widget class="QLabel" name="label_12">
+                   <property name="sizePolicy">
+                    <sizepolicy hsizetype="Expanding" vsizetype="Preferred">
+                     <horstretch>0</horstretch>
+                     <verstretch>0</verstretch>
+                    </sizepolicy>
+                   </property>
+                   <property name="text">
+                    <string comment="After 10 MB backup the log file">the log file</string>
+                   </property>
+                  </widget>
+                 </item>
+                </layout>
                </item>
               </layout>
              </widget>
```

sledgehammer999 commented:
Rebased and applied your observations. Start reviewing from the 6th commit onwards. I did separate commits to ease your review. Those commits will be squashed into their respective first 5 commits.

sledgehammer999 commented:
**Rant:** I wish Github ordered the commits by commit date and not author date. I first randomly commited stuff and then rearranged the commits, but it doesn't show in this history. Luckily git's history sorts by commit date

sledgehammer999 commented:
(fixed travis)

sledgehammer999 commented:
@glassez tell me if it is ok to merge when you're ready.

glassez commented on src/app/application.cpp:
```diff
@@ -465,6 +477,7 @@ void Application::cleanup()
     Net::DownloadManager::freeInstance();
     Preferences::freeInstance();
     Logger::freeInstance();
+    delete m_fileLogger;
```
Move up. Before `Logger::freeInstance();` since FileLogger uses Logger.

sledgehammer999 replied:
Yes, but it never accesses logger directly, only in construction to setup the signal/slots. My intention was to leave it last in order to catch all messages to file.
Do you still want to move it?

glassez replied:
I don't think you can lose some messages here. Better change it.

sledgehammer999 replied:
ok

glassez commented on src/app/application.cpp:
```diff
@@ -101,10 +102,21 @@ Application::Application(const QString &id, int &argc, char **argv)
 
+void Application::configure()
+{
+    m_fileLogEnabled = Preferences::instance()->fileLogEnabled();
```
`m_fileLogEnabled` is redundant variable.

sledgehammer999 replied:
true. I'll remove.

glassez commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,107 @@
+    configure();
+    const Logger* const logger = Logger::instance();
+    foreach (const Log::Msg& msg, logger->getMessages())
+        addLogMessage(msg);
```
Here you open/write/close log file several times in a row. I think that is wasteful.

sledgehammer999 replied:
I was thinking this when I was implementing it. I was going to discuss about this but I forgot.
My other thought was: Have the file as member var and keep it open for eg 5secs of inactivity(timer). Then close it again.
Or keep it open indefinitely and close on object destruction?

glassez replied:
> keep it open indefinitely and close on object destruction?

Yes. And do flush() from time to time.

sledgehammer999 replied:
ok another timer... What timeout do you suggest?

glassez replied:
Maybe just to use 0 timeout? Do you understand what I mean? If timer isn't active we call single shot with 0 timeout in addLogMessage and store that timer is active.

sledgehammer999 replied:
So if the control doesn't return to mainloop we continue to output stuff to the file. But when control returns to mainloop it means that it is unlikely to have other stuff to write immediately, the slot is called and the file is closed.
Nice idea, but IMO we will have almost the same problem with the warning messages about the trackers.(when retrieving favicons that don't exist)

glassez replied:
>we will have almost the same problem with the warning messages about the trackers.(when retrieving favicons that don't exist)

What's the problem you're talking about?

sledgehammer999 replied:
When application is started it tries to download the favicons for the torrents' trackers. There are quite a few trackers that don't return a favicon and we output warning messages to the log. Just check your log to see the orange text.
These messages are being logged one after the other at the first few second(or minute) of startup.
So I have timer with a 0 timeout to close the filelog file: The file will be opened again after the next mainloop iteration. Then closed again and then open agian etc. So we are back to the original problem.

glassez replied:
No. I think it's a problem of different level. Here you are trying to do this in a loop (with delays in a few micro or even nano seconds). And in another case the delay is much more. Besides the messages about favicons not so much (theoretically, their may not be). I also suggest not to open/close the file and keep it open and periodically do flush().

sledgehammer999 replied:
>periodically do flush().

I forgot this. I had in my mind close(). Then I am ok.

glassez commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,107 @@
+
+    QTextStream str(&log);
+    QDateTime time = QDateTime::fromMSecsSinceEpoch(msg.timestamp);
+    QString type;
```
Maybe it would be better to use string array indexed by message type?

sledgehammer999 replied:
like we did recently with the SIGTERM etc signals?
That is a good thought.

glassez commented on src/base/logger.h:
```diff
@@ -6,17 +6,19 @@
+        CRITICAL = 0x8 //ERROR is defined by libtorrent and results in compiler error
     };
+    Q_DECLARE_FLAGS(MsgTypes, MsgType)
+    Q_DECLARE_OPERATORS_FOR_FLAGS(Log::MsgTypes)
```
Isn't `Q_DECLARE_OPERATORS_FOR_FLAGS` supposed to be placed outside the `Log` namespace?

sledgehammer999 replied:
I have no idea... The docs aren't very detailed there. Plus it seems to work on the other places where I use bitwise operations.
But I can move it if you want.

glassez replied:
I can't read docs at this time but I think it is clear about this.

evsh replied:
You better read its source code ;) It has to be placed inside of the namespace for C++ overload rules to work as expected. And `Log::` can be omitted, I guess.

glassez replied:
Qt Example:
```c++
class MyClass
{
public:
    enum Option {
        NoOptions = 0x0,
        ShowTabs = 0x1,
        ShowAll = 0x2,
        SqueezeBlank = 0x4
    };
    Q_DECLARE_FLAGS(Options, Option)
    ...
};

Q_DECLARE_OPERATORS_FOR_FLAGS(MyClass::Options)
```

glassez replied:
Q_DECLARE_FLAGS is inside namespace.
Q_DECLARE_OPERATORS_FOR_FLAGS is outside namespace (but it uses full qualified type name).

evsh replied:
Fine, but where is a namespace?

glassez replied:
You disappoint me... What a class is different from the namespace in this context?

evsh replied:
For operators the lookup mechanism checks for an user-provided version in the namespaces where its arguments are declared. Therefore there is no need to pollute the global namespace.

glassez replied:
Ah, now I see. I'm sorry, I didn't know about it (apparently, Qt developers too).
For followers: [Argument-dependent lookup](http://en.cppreference.com/w/cpp/language/adl)

glassez replied:
Besides, there will be no problem, if you leave it as is.
>Therefore there is no need to pollute the global namespace.

Says it's too loud. This won't hurt it.

glassez commented on src/gui/executionlog.h:
```diff
@@ -32,27 +32,22 @@
 #define EXECUTIONLOG_H
 
 #include <QWidget>
+#include <base/logger.h>
```
It is better to divide the system and local includes. Use `"` with local.

sledgehammer999 replied:
Sorry. My mistake.

glassez commented on src/gui/loglistwidget.cpp:
```diff
@@ -37,60 +37,68 @@
-  QListWidget(parent),
-  m_maxLines(max_lines)
+LogListWidget::LogListWidget(int maxLines, const Log::MsgTypes &types, QWidget *parent) :
+    QListWidget(parent),
```
Coding Style!!!

sledgehammer999 replied:
I feel ashamed.

glassez commented on src/gui/loglistwidget.cpp:
```diff
@@ -37,60 +37,68 @@
-  clear();
+void LogListWidget::copySelection()
+{
+    static QRegExp html_tag("<[^>]+>");
```
camel Case

sledgehammer999 replied:
not my code but ok.

glassez commented on src/gui/loglistwidget.cpp:
```diff
@@ -37,60 +37,68 @@
+void LogListWidget::copySelection()
+{
+    static QRegExp html_tag("<[^>]+>");
+    QList<QListWidgetItem*> items = selectedItems();
```
This variable is redundant.

sledgehammer999 replied:
:+1:

glassez commented on src/base/logger.h:
```diff
@@ -6,17 +6,19 @@
+        NORMAL = 0x1,
+        INFO = 0x2,
+        WARNING = 0x4,
+        CRITICAL = 0x8 //ERROR is defined by libtorrent and results in compiler error
```
Maybe add `ALL = -1` here and use it later?

sledgehammer999 replied:
OK. Should I put it first or last?

glassez replied:
No matter.

glassez commented:
@sledgehammer999 I finished the review.

glassez commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,107 @@
+FileLogger::FileLogger()
+{
+    configure();
+    const Logger* const logger = Logger::instance();
```
Shouldn't we start from new file here?

sledgehammer999 replied:
I don't follow. What do you mean?

glassez replied:
Imagine this situation: when we run the program, we have some log file (almost complete), remaining after the previous session. We adding some number of messages, it fills up, we remove it and start writing to a new file. We may lose some interesting messages, for example, torrents resuming, ports allocation, etc.

sledgehammer999 replied:
To be honest, I have seen the GUI options other apps have that allow delete/backup for X size. But I have never used them to actually know how they handle that situation(your use case).
Do you have any proposal here?
And isn't it a bit dangerous anyway to select "delete after X size"? I mean the user probably accepts the risk.
PS: I though of an idea:
1. Allow to only backup after X size and
2. Allow to delete **backup** logs that are X days old?

glassez replied:
>Do you have any proposal here?

At least start writing to new file when we start qBittorrent. And let the rest up to the user.

glassez replied:
>Allow to delete backup logs that are X days old

It's good idea. In addition to corresponding option.

sledgehammer999 replied:
>At least start writing to new file when we start qBittorrent. And let the rest up to the user.

No. That will fill up the log folder with too much small log files in very short time.

glassez replied:
I mean case when we delete files. We just delete the old file and starts writing to new one.

sledgehammer999 replied:
Well there are a lot of bugs reported where older log entries would have helped in faster debugging.
Anyway, I'll implement the delete after X days and be done with it.

LordNyriox commented:
@sledgehammer999:  If you're still planning on merging this in, you should probably rebase.  

Thanks, 
~ Nyriox

sledgehammer999 commented:
>@sledgehammer999: If you're still planning on merging this in, you should probably rebase. 

I know you're trying to help but this comment was unnecessary. I am both the PR author and the one that will merge. So I'll know I need rebasing.

evsh commented on src/app/application.h:
```diff
@@ -103,6 +106,10 @@ private slots:
 
+    // FileLog
+    bool m_fileLogEnabled;
+    QPointer<FileLogger> m_fileLogger;
```
Why not  `QScoopedPointer`? The `Application` object owns the `m_fileLogger` object, doesn't it?

sledgehammer999 replied:
We do cleanup in Application::cleanup() instead of destructor. The reason is we need to control the shutdown process on Windows otherwise windows kills us aka unclean exit.

evsh replied:
I see. Thank you for the reply.

On 13 March 2016 at 16:38, sledgehammer999 <notifications@github.com> wrote:

> In src/app/application.h
> <https://github.com/qbittorrent/qBittorrent/pull/4681#discussion_r55938801>
> :
>
> > @@ -103,6 +106,10 @@ private slots:
> >      QPointer<WebUI> m_webui;
> >  #endif
> >
> > +    // FileLog
> > +    bool m_fileLogEnabled;
> > +    QPointer<FileLogger> m_fileLogger;
>
> We do cleanup in Application::cleanup() instead of destructor. The reason
> is we need to control the shutdown process on Windows otherwise windows
> kills us aka unclean exit.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/qbittorrent/qBittorrent/pull/4681/files#r55938801>.
>

evsh commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,107 @@
+
+    str << type << time.toString(Qt::ISODate) << " - " << msg.message << endl;
+
+    if (m_manage && (log.size() >= (m_size * 1024 * 1024))) {
```
Since you've put file size check after writing the current message, a `str.flush()` call would be logical in between.

evsh commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,107 @@
+    if (m_manage && (log.size() >= (m_size * 1024 * 1024))) {
+        log.close();
+
+        if (m_fileLogAction == REMOVE) {
```
I would move the tests and the file manipulation towards the beginning of the function, before the place where the current message is written. Then the call to `flush()`, mentioned above is not needed, of course.

glassez replied:
But we have to add a message to the file, and only then check whether it has exceeded the limit, and decide what to do with it.

evsh replied:
So what can be wrong if we chech that before the next write?

glassez replied:
Hmm... Maybe you're right.

evsh commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,107 @@
+        m_path += "/";
+    QDir dir(m_path);
+    if (!dir.exists())
+        dir.mkpath(m_path);
```
And `mkpath()` will again check does the path exist.

evsh commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,107 @@
+    QDir dir(m_path);
+    if (!dir.exists())
+        dir.mkpath(m_path);
+    m_path += "qbittorrent.log";
```
I would write here
 ```C++
m_path = dir.absoluteFilePath("qbittorrent.log")
```
and remove `endsWith()` check above.

evsh commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,107 @@
+        return;
+
+    QTextStream str(&log);
+    QDateTime time = QDateTime::fromMSecsSinceEpoch(msg.timestamp);
```
`time` is redundant variable unless you want to debug the `QDateTime::fromMSecsSinceEpoch()` call.

evsh commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,107 @@
+{
+    QFile log(m_path);
+    if (!log.open(QIODevice::WriteOnly | QIODevice::Append | QIODevice::Text))
+        return;
```
Perhaps it is possible to write into the log window a note about this fiasco?

glassez replied:
And get an infinite loop here?

evsh replied:
That is why I wrote "perhaps"

evsh replied:
So, I would add a flag to the `FileLogger` class, say `m_ignoreMessages`, set it here, fire a log message and in the very beginning of the function I'll check:
```C++
if (m_ignoreMessages)
    return;
```
And reset the flag in `FileLogger::configure()` after everything is set up. As a bonus `FileLogger` will become almost ready for multi-thread usage.

glassez replied:
Are you going again across the universe... Who (or what) will set this flag? FileLogger is a Logger client, do not forget about it.
But something to do is still possible... for Example, the FileLogger can ignore messages on any (hardcoded) attribute.

evsh replied:
This function can set it directly here:
```C++
if (!log.open(QIODevice::WriteOnly | QIODevice::Append | QIODevice::Text)) {
    m_ignoreMessages = true;
    sendWarningMessageToTheLog(logFileOpenOperationHasFailedMessage());
    return;
}
```
Also, if `Utils::Fs::forceRemove()` call below fails, we can do the similar things.

glassez replied:
It makes sense (we can safely assume that if we are unable to open the file once, we will not be able to do it next). Although it is better to name the flag something more suitable (`m_invalid` or something similar).

glassez replied:
But better to release it first without this feature.

evsh commented on src/base/preferences.cpp:
```diff
@@ -62,6 +62,8 @@
 
 Preferences* Preferences::m_instance = 0;
 
+static const QString LOG_FOLDER("logs");
```
Just want to note that the usage of `static` is deprecated for this purpose by the C++11 standard. Since the code is new, perhaps is might make sense to use the anonymous namespace.

evsh commented on src/base/preferences.cpp:
```diff
@@ -1054,6 +1056,87 @@ void Preferences::setExecutionLogEnabled(bool b)
+    int val = value("Preferences/FileLog/MaxSize", 10).toInt();
+    if (val < 1)
+        return 1;
+    else if (val > 1000)
```
`else` is redundant

Chocobo1 replied:
or just use `min()` & `max()` to clamp it's value, just a matter of preference...

glassez replied:
>else is redundant

And how are you going to return a value between 1 and 1000?

evsh replied:
See below.

evsh replied:
I mean that only the keyword `else` is redundant, but not the `else` clause.

glassez replied:
I don't understand what do you mean.
We have three cases:
val <= 1: we should return 1
1 < val <= 1000: we should return val itself
val > 1000: we should return 1000
and we have three branches in code which corresponds to three cases above. What is redundant?

evsh replied:
```C++
if (val < 1)
    return 1;
if (val > 1000)
    return 1000;
return val;
```
You don't put `break` after `return` in `switch`, right? The same is here.

glassez replied:
>I mean that only the keyword else is redundant, but not the else clause.

You should be more specific. Then other explanations need not.

evsh commented on src/base/preferences.cpp:
```diff
@@ -1054,6 +1056,87 @@ void Preferences::setExecutionLogEnabled(bool b)
+    if (val < 0 || val > 1)
+        return 0;
+    else
+        return val;
```
```C++
return (val < 0 || val > 1) ? 0 : val;
```

evsh commented on src/base/preferences.cpp:
```diff
@@ -1054,6 +1056,87 @@ void Preferences::setExecutionLogEnabled(bool b)
+
+void Preferences::setFileLogAction(const int &action)
+{
+    if (action < 0 || action > 1)
```
```C++
setValue("Preferences/FileLog/Action", action < 0 || action > 1 ? 0: action)
```
is more explicit

evsh commented on src/base/preferences.cpp:
```diff
@@ -1054,6 +1056,87 @@ void Preferences::setExecutionLogEnabled(bool b)
+
+void Preferences::setFileLogMaxSize(const int &size)
+{
+    if (size < 1)
```
Questionable, but...
```C++
setValue("Preferences/FileLog/MaxSize", size < 1 ? 1 : size > 1000 ? 1000 : size);
```
or
```C++
setValue("Preferences/FileLog/MaxSize", std::min(std::max(size, 1), 1000));
```

glassez replied:
```c++
setValue("Preferences/FileLog/MaxSize", size < 1 ? 1 : size > 1000 ? 1000 : size);
```
It's unreadable.

```c++
setValue("Preferences/FileLog/MaxSize", std::min(std::max(size, 1), 1000));
```
It's better.

glassez commented on src/gui/executionlog.cpp:
```diff
@@ -35,18 +35,20 @@
 #include "executionlog.h"
 #include "ui_executionlog.h"
-#include "base/logger.h"
+#include "base/preferences.h"
```
Recall that **Preferences** is deprecated in new code.

evsh commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,107 @@
+void FileLogger::addLogMessage(const Log::Msg &msg)
+{
+    QFile log(m_path);
+    if (!log.open(QIODevice::WriteOnly | QIODevice::Append | QIODevice::Text))
```
Since there is a default value for the log file name, I would like to see here a call to `QFile::setPermissions(QFile::ReadOwner|QFile::WriteOwner)`

evsh commented:
@sledgehammer999: I just wonder why file log here is custom implemented instead of, say, boost::logging (since libtorrent anyway gives us boost as a dependency). Is is a project policy or there is another reason behind this decision?

sledgehammer999 commented:
I pushed an updated version. I implemented most of the stuff you talked about. Start reviewing from commit ` 	Remove redundant variable.` onwards.
Pretty much all commits will be squashed before final merge.

Here is how the new options look:
![untitled](https://cloud.githubusercontent.com/assets/273315/13730908/a2755eee-e965-11e5-9aec-135c2930dca0.jpg)

sledgehammer999 commented:
>@sledgehammer999: I just wonder why file log here is custom implemented instead of, say, boost::logging (since libtorrent anyway gives us boost as a dependency). Is is a project policy or there is another reason behind this decision?

Actually I didn't investigate on other logging libraries. I wanted something custom and "lightweight". I assume other libs implement a ton of stuff that we don't need.
There isn't a project policy that discourages using other libs. But not for trivial stuff. eg we use QJson for qt4 builds.

LordNyriox commented:
@sledgehammer999:  I believe @glassez would point out that your backup/delete feature uses the `preferences` class which he apparently deprecated in commit [https://github.com/qbittorrent/qBittorrent/commit/d05d5a85a506b9f7a76d5d74df953e931aa827af] during development of the ASM feature.  

I don't think I'm qualified for a deeper review than that.  

Thanks, 
~ Nyriox

glassez commented:
>@sledgehammer999: I believe @glassez would point out that your backup/delete feature uses the  preferences  class which he apparently deprecated in commit [ d05d5a8 ] during development of the ASM feature.

Sure. There is my comment on it here.
As I said, I don't insist anyone to fix existing code, but we should not use this in new code (I mean, we don't have to add anything to Preferences).

sledgehammer999 commented:
@glassez I pushed a new commit using Settings Storage.
I am not sure I have done it correctly. Also I need your help. I am lost. I don't know how to migrate the `FileLogger::configure()` and `Application::configure()` which reacted to changed Preferences to the new system. I have migrated some of it, but currently it **will not react**:
1. If user disables file logging
2. If user changes log path
3. If user enables/disables "delete old backup" files or changes the age to delete them.

LordNyriox commented:
@sledgehammer999:  Ugh.  I may be a code klutz, but even *I* know better than that.  If you examine the commit mentioned, the `settingsstorage` change is completely transparent.  

You literally need to revert that whole last commit and simply replace any `preferences.h` includes with `settingsstorage.h` includes.  

All other `pref` and `preferences` stuff is left the same.  It's simply the file for the backend code that is different.  

Thanks,
~ Nyriox

LordNyriox commented:
@glassez:  Please rescue @sledgehammer999 from this mess. 

Thanks, 
~ Nyriox

sledgehammer999 commented:
>@glassez: Please rescue @sledgehammer999 from this mess. 

Please stop doing that. He is already notified by my earlier comment.
Also don't delete your comments when someone quotes them or responds to you. You make him like a crazy person talking at the wind.

LordNyriox commented:
@sledgehammer999:  The issue here is that I make a lot of stupid comments due to my ignorance of anything code-related, so if I feel if I say something, after it's been responded to it would be better to make my stupidity disappear (at least eventually).  If you'd prefer, I can wait until an issue is closed (or merged) before deleting my stupid comments, but I would like to delete them eventually if at all possible.  

Thank you for your patience dealing with me. 
~ Nyriox

glassez commented:
>I am not sure I have done it correctly.

You're right. You done it incorrectly.
Apparently, I need to describe the principles of the new code:
1. Components owns its properties and stores/loads needed data from SettingsStorage itself. The main thing is to understand which property relates to a component (this is not always evident at first sight).
2. No `configure()` slot in new classes. All things we need to do when properties changed we should do in appropriate property setters.
3. Properties should be implemented as setter/getter pair (or only as getter if property is read only). Setter/Getter are usually non-static methods (AddNewTorrentDialog is bad example so it's exception from this rules).
4. If some component needs to store some private data in SettingsStorage (e.g. its state) there are no getter/setter for this data (or some private helper methods if needed).
5. All non-toplevel components should be accessible as properties of toplevel components (e.g. Application::fileLogger()) so we can get/set its properties. If some component is non full time existed (as FileLogger) we need check its existence before access its properties.

I've implemented AddNewTorrentDialog::isEnabled() (and other property) as static method because we really do nothing when we change these properties. But this way is incorrect for FileLogger. Really we haven't FileLogger::isEnabled property but we have Application::isFileLoggerEnabled. Example:
```c++
void Application::setFileLoggerEnabled(bool enabled)
{
    if (enabled) {
        if (!m_fileLogger) {
            m_fileLogger = new FileLogger;
            // store in settings here
        }
    }
    else {
        if (m_fileLogger) {
            delete m_fileLogger;
            // store in settings here
        }
    }
}
```

@sledgehammer999 I hope I have covered it fully. But you are free to ask more questions.

sledgehammer999 commented:
Yes, I think I am ok with your explanation. As you guessed it I took AddNewTorrentDialog as an example.
I apparently need to implement this one level up.
FileLogger goes to Application and ExecutionLog to MainWindow.
I'll give another try in a few moments.

sledgehammer999 commented:
Hmm, I was going to start but I already hit a wall. `FileLogger` properties need to be loaded/stored from options_imp.cpp too. How do I do that? Do I need to pass an `Application` pointer to options_imp ?

glassez commented:
Application is Singleton. You can access it via `static_cast<Application*>(QCoreApplication::instance())`.

evsh commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,267 @@
+    if (!m_logFile) return;
+
+    QTextStream str(m_logFile);
+    QString type;
```
The variable may be removed, perhaps. `str << "(I) "` does not look worse than  `type = "(I) "`

sledgehammer999 replied:
I ignored this, but ok I'll do it.

evsh commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,267 @@
+
+    str << type << QDateTime::fromMSecsSinceEpoch(msg.timestamp).toString(Qt::ISODate) << " - " << msg.message << endl;
+
+    if (isBackup() && (m_logFile->size() >= (maxSize() * 1024 * 1024))) {
```
Shouldn't one call `str.flush()` here or delete the `str` object?

sledgehammer999 replied:
Hmm, you're probably right. I'll fix it.

sledgehammer999 replied:
According to docs `endl` flushes the stream too.

evsh commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,267 @@
+        || !m_logFile->setPermissions(QFile::ReadOwner | QFile::WriteOwner)) {
+        delete m_logFile;
+        m_logFile = nullptr;
+        Logger::instance()->addMessage(tr("An error occured while trying to open the filelog. Logging to file is disabled."), Log::CRITICAL);
```
'filelog'? a typo, perhaps?

sledgehammer999 replied:
No. I want to indicate the file on disk. Do you have a better word?

LordNyriox replied:
@sledgehammer999:  Try `log-file` or `log file` instead.  It is more gramatically correct.

evsh replied:
Agree with @LordNyriox , "log file" is perfectly fine.

evsh commented on src/gui/loglistwidget.h:
```diff
@@ -31,31 +31,33 @@
 
 public slots:
-  void appendLine(const QString &line);
+    void appendLine(const QString &line, const Log::MsgType &type);
```
Why I'm not allowed to indent with tabs in this project? I like them so much!

sledgehammer999 replied:
Because we have decided a long ago in 4-spaces indentation.

evsh replied:
Yes, I know, of course. Was joking after I've seen tab indentation in your code.

sledgehammer999 replied:
Nope. That's 2 more spaces to make it 4 spaces in total.

evsh replied:
Oh, I was wrong. Excuse me my stupid joke, please.

sledgehammer999 commented:
>Application is Singleton. You can access it via static_cast<Application*>(QCoreApplication::instance()).

OK.
And that means that options_imp will also have to have access to the mainwindow pointer(for the other unported settings) , right?

glassez commented:
>And that means that options_imp will also have to have access to the mainwindow pointer(for the other unported settings) , right?

Right.
It is accessible via Application.

sledgehammer999 commented:
I squashed commits. I implemented the suggestions and rebased. The convertion to SettingsStorage will remain in a separate commit, because I want all the previous ones to include them in v3.3.4.
SettingsStorage along with ASM will be released in v3.3.5.

@glassez can you review my new attempt at SettingsStorage?
I have realized a possible "bug zone" for bigger components. As you can see in options_imp I reordered how the settings are saved for FileLogger, otherwise the setters wouldn't work properly. eg If I first set the enable property and then the backup property filelogger will use the old backup value for a brief moment.
I assume that converting the Session class will need extra care for these subtle bugs.

I also decided to have each component live under its own namespace(?) in the text file without mentioning the parent that they belong in. For example filelogger saves settings under `FileLogger/` path and not `Application/FileLogger/`. Is that ok?

Chocobo1 commented on src/app/application.h:
```diff
@@ -74,6 +76,22 @@ class Application : public BaseApplication
+    bool isFileLoggerDeleteOld() const;
+    void setFileLoggerDeleteOld(bool value);
+    int getFileLoggerMaxSize() const;
+    void setFileLoggerMaxSize(const int &value);
```
maybe just pass-by-value? (for simple types)

glassez replied:
This always surprises me too.

sledgehammer999 replied:
I can fix this. Is this the only thing about the PR?

glassez replied:
I haven't reviewed it after your last changes. I'll do it ASAP.

evsh commented on src/app/application.cpp:
```diff
@@ -106,9 +123,105 @@ Application::Application(const QString &id, int &argc, char **argv)
+    if (value && !m_fileLogger)
+        m_fileLogger = new FileLogger(getFileLoggerPath(), isFileLoggerBackup(), getFileLoggerMaxSize(), isFileLoggerDeleteOld(), getFileLoggerAge(), static_cast<FileLogger::FileLogAgeType>(getFileLoggerAgeType()));
+    else
+        delete m_fileLogger;
```
What if `value == true` but `m_fileLogger != nullptr`?

glassez replied:
+1
`else if (!value)`

sledgehammer999 replied:
Oops. I am going to fix it.

evsh commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,179 @@
+
+void FileLogger::changePath(const QString& newPath)
+{
+    QString tmpPath = Utils::Fs::fromNativePath(newPath);
```
What this black magic is all about? Why not just:
```C++
QDir dir (tmpPath);
dir.mkpath(tmpPath);
QString tmpPath = Utils::Fs::fromNativePath(dir.absoluteFilePath("qbittorrent.log"));
```
?

glassez replied:
+1
But `newPath` in first and second statements.

evsh replied:
@glassez: thank you, I'll fix my comment above.

sledgehammer999 replied:
I am not sure what you guys are talking about. There is an agreed rule to use internally paths with "/" instead of "\" even if the called functions understand both(like QDir). Or are you talking about something else entirely?

evsh replied:
I'm talking about lines 66 – 71 (6 in total). They can be replaced with 3 lines. And I do not understand what is is reason behind lines 67 and 68.

sledgehammer999 replied:
Line 66 makes the path look as "c:/test" instead of "c:\test". The user might have edited the ini file and put the path with "\" characters. So I am enforcing here the rule I talked about previously.
Lines 67-68 append missing "/" because at line 71 we append the filename.

evsh replied:
Thank you, but why to do that manually with two lines of code, if you can invoke `QDir::absoluteFilePath()`?

sledgehammer999 replied:
Ok, I'll drop the if block and use QDir::absoluteFilePath() instead.

glassez commented on src/app/application.cpp:
```diff
@@ -106,18 +122,104 @@ Application::Application(const QString &id, int &argc, char **argv)
+
+QString Application::getFileLoggerPath() const
+{
+    return settings()->loadValue(KEY_FILELOGGER_PATH, QVariant(Utils::Fs::QDesktopServicesDataLocation() + LOG_FOLDER)).toString();
```
Just `Utils::Fs::QDesktopServicesDataLocation() + LOG_FOLDER` isn't enough here?

glassez commented on src/app/filelogger.cpp:
```diff
@@ -87,32 +80,41 @@ void FileLogger::configure()
-        switch (static_cast<FileLogAgeType>(pref->fileLogAgeType())) {
-        case DAYS:
-            date = date.addDays(pref->fileLogAge());
+    foreach (const QFileInfo file, dir.entryInfoList(QStringList() << "qbittorrent.log.bak*", QDir::Files | QDir::Writable, QDir::Time | QDir::Reversed)) {
```
Can you try `{ "qbittorrent.log.bak*" }` instead of `QStringList() << "qbittorrent.log.bak*"`?

glassez commented on src/gui/executionlog.h:
```diff
@@ -46,7 +46,7 @@ class ExecutionLog: public QWidget
 
 public:
-    explicit ExecutionLog(QWidget *parent = 0);
+    explicit ExecutionLog(QWidget *parent, const Log::MsgTypes &types);
```
`explicit` isn't needed.

glassez commented on src/gui/mainwindow.h:
```diff
@@ -81,6 +81,12 @@ class MainWindow: public QMainWindow, private Ui::MainWindow
+    // ExecutionLog properties
+    bool isExecutionLogEnabled() const;
+    void setExecutionLogEnabled(bool value);
+    int getExecutionLogMsgTypes() const;
```
Why do you use non-Qt naming? Just `executionLogMsgTypes` as we usually did. This affects all this PR.

sledgehammer999 replied:
ok

glassez commented:
>I have realized a possible "bug zone" for bigger components. As you can see in options_imp I reordered how the settings are saved for FileLogger, otherwise the setters wouldn't work properly. eg If I first set the enable property and then the backup property filelogger will use the old backup value for a brief moment.

Everything is logical and it always has been, just in a different place. Just it looked not so clearly. I don't think it will give us problems.
>I also decided to have each component live under its own namespace(?) in the text file without mentioning the parent that they belong in. For example filelogger saves settings under FileLogger/ path and not Application/FileLogger/. Is that ok?

In this case, we do not have the FileLogger settings, only the Application settings. So consider using names like "Application/FileLoggerEnabled" and so on. It turns out that we do not have a standalone FileLogger component.

glassez commented on src/app/application.cpp:
```diff
@@ -71,7 +72,23 @@
+namespace
+{
+#define FILELOGGER_SETTINGS_KEY(name) "FileLogger/" name
+    const QString KEY_FILELOGGER_ENABLED = FILELOGGER_SETTINGS_KEY("Enabled");
```
It is not necessary to put constants in an anonymous namespace.

evsh replied:
But it saves one from typing 'static' in each line.

glassez replied:
It is not necessary to prepend constants with `static`. Constants use internal linkage by default.

sledgehammer999 replied:
But since I am creating `inline SettingsStorage *settings()` I opted to put everything here too to reduce visual clutter.
I am leaving it as is.

evsh replied:
Thank you!

sledgehammer999 replied:
>@glassez : Thank you!

!?!?!

If you guys think it is better, I can leave them out of the namespace...

glassez replied:
>@glassez : Thank you!

We share with @evsh the missing knowledge :) Apparently, it's for this.

evsh replied:
Yes

sledgehammer999 replied:
So is it ok to leave all this inside the namespace for visual clarity?

evsh replied:
@sledgehammer999 , I don't think that without namespace the code will be better. I, personally, like anonymous namespace because they make intents exceptionally clear. One will not think whether there are any `extern` declarations, declarations in a header file, or so.

sledgehammer999 replied:
OK. I think we're wasting time talking more about this.

glassez replied:
>So is it ok to leave all this inside the namespace for visual clarity?

Yes. But `#define` looks ugly here. Anyway, it caught my eyes initially.

glassez commented on src/app/application.cpp:
```diff
@@ -106,9 +123,105 @@ Application::Application(const QString &id, int &argc, char **argv)
+
+QString Application::getFileLoggerPath() const
+{
+    return settings()->loadValue(KEY_FILELOGGER_PATH, QVariant(Utils::Fs::QDesktopServicesDataLocation() + LOG_FOLDER)).toString();
```
Just `Utils::Fs::QDesktopServicesDataLocation() + LOG_FOLDER` isn't enough here?

sledgehammer999 replied:
+1 I think I just blindly copied it from something else in Preferences.

sledgehammer999 replied:
I think I had a dirty build dir before. Now a clean build give me this error:
```
src\app\application.cpp:148: error: C2664: 'QVariant SettingsStorage::loadValue(const QString &,const QVariant &) const': cannot convert argument 2 from 'QStringBuilder<QString,QString>' to 'const QVariant &'
src\app\application.cpp:148: error: C2228: left of '.toString' must have class/struct/union
```

glassez replied:
Weird. What does `QStringBuilder<QString,QString>`? `operator+(const QString &, const QString &)` must return `const QString()` according to the documentation.

sledgehammer999 replied:
I don't know :S

glassez commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,179 @@
+        date = date.addYears(age);
+    }
+
+    foreach (const QFileInfo file, dir.entryInfoList(QStringList() << "qbittorrent.log.bak*", QDir::Files | QDir::Writable, QDir::Time | QDir::Reversed)) {
```
Can you try `{ "qbittorrent.log.bak*" }` instead of `QStringList() << "qbittorrent.log.bak*"`?

sledgehammer999 replied:
I changed it to:
```c++
foreach (const QFileInfo file, dir.entryInfoList({ "qbittorrent.log.bak*" }, QDir::Files | QDir::Writable, QDir::Time | QDir::Reversed)) {
        if (file.lastModified() < date)
            break;
        Utils::Fs::forceRemove(file.absoluteFilePath());
    }
```

But I get:
>src\app\filelogger.cpp:101: error: C1075: the left brace '{' was unmatched at the end of the file

glassez replied:
Weird... Then you can leave it as is or make it slightly shorter: `QStringList("qbittorrent.log.bak*")`.

glassez commented on src/app/filelogger.cpp:
```diff
@@ -0,0 +1,179 @@
+}
+
+void FileLogger::openLogFile()
+{
```
Wouldn't it be easier if you put the file creation here, and deleting in `closeLogFile()`?
I know that in one place it will lead to we create it immediately after removal. For me, it's a small problem.

sledgehammer999 replied:
I think it is cleaner this way, so I am leaving it.

glassez commented on src/base/logger.h:
```diff
@@ -6,17 +6,19 @@
+        NORMAL = 0x1,
+        INFO = 0x2,
+        WARNING = 0x4,
+        CRITICAL = 0x8 //ERROR is defined by libtorrent and results in compiler error
```
`Log::ERROR` is defined by libtorrent?

sledgehammer999 replied:
I thought you were right and this was a leftover from my initial implementation. But after changing it I get these cryptic messages from msvc:
```
src\base\logger.h:19: error: C2059: syntax error: 'constant'
src\base\logger.h:19: error: C3805: 'constant': unexpected token, expected either '}' or a ','
```

glassez replied:
Apparently, the constant from libtorrent still interferes here.

glassez commented on src/gui/executionlog.cpp:
```diff
@@ -35,18 +35,19 @@
     ui->setupUi(this);
 
+    m_msgList = new LogListWidget(MAX_LOG_MESSAGES,
+                                  Log::MsgTypes(types));
```
Why not in one line? It isn't so long.

sledgehammer999 replied:
OK. It was something else before. I am changing it.

glassez commented on src/gui/executionlog.h:
```diff
@@ -32,27 +32,22 @@
 
 public:
-    explicit ExecutionLog(QWidget *parent = 0);
+    explicit ExecutionLog(QWidget *parent, const Log::MsgTypes &types);
```
explicit isn't needed.

I'm sorry. Mobile GitHub is not working lately. All my comments are added in the wrong place.

sledgehammer999 replied:
OK, I'll remove it.

glassez commented on src/gui/loglistwidget.h:
```diff
@@ -31,31 +31,33 @@
 public:
-  explicit LogListWidget(int max_lines = 100, QWidget *parent = 0);
+    // -1 is the portable way to have all the bits set
+    explicit LogListWidget(int maxLines, const Log::MsgTypes &types = Log::MsgTypes(-1), QWidget *parent = 0);
```
Log::ALL?

sledgehammer999 replied:
Oops. That was before I added the Log::ALL.
PS: I should remove the `explicit` keyowrd too, right?

glassez replied:
> I should remove the explicit keyowrd too, right?

No. Not here, because we have the ability to call the constructor with one parameter.

glassez commented:
@sledgehammer999 I finished with this PR.

sledgehammer999 commented:
>In this case, we do not have the FileLogger settings, only the Application settings. So consider using names like "Application/FileLoggerEnabled" and so on. It turns out that we do not have a standalone FileLogger component

But won't this become cumbersome? I mean, eg MainWindow might turn up with very long options names. Or depending on the option it may look too big(?) eg "Application/FileLoggerAgeType" or "Application/FileLoggerDeleteOld"

glassez commented:
>But won't this become cumbersome? I mean, eg MainWindow might turn up with very long options names. Or depending on the option it may look too big(?) eg "Application/FileLoggerAgeType" or "Application/FileLoggerDeleteOld"

Ok. Let it be as you suggested.
But why are you against nested names?

sledgehammer999 commented:
>But why are you against nested names?

I don't have a solid reason. I just feel that it is more manageable without it. However, I like much more `Application/FileLogger/Enabled` instead of `Application/FileLoggerEnabled`.

I'll reconsider nested names, because I realized once the SettingsStorage convertion is complete we might end up with a shitload of small modules in the ini files. Not manageable at all.

glassez commented:
>I'll reconsider nested names, because I realized once the SettingsStorage convertion is complete we might end up with a shitload of small modules in the ini files. Not manageable at all.

+1

sledgehammer999 commented:
I think I'll switch to nested names. Would the following be considered ok or does it constitute preprocessor abuse?
```c++
namespace
{
#define SETTINGS_KEY(name) "Application/" name

    // FileLogger properties keys
#define FILELOGGER_SETTINGS_KEY(name) SETTINGS_KEY("FileLogger/") name
    const QString KEY_FILELOGGER_ENABLED = FILELOGGER_SETTINGS_KEY("Enabled");
    const QString KEY_FILELOGGER_PATH = FILELOGGER_SETTINGS_KEY("Path");
    const QString KEY_FILELOGGER_BACKUP = FILELOGGER_SETTINGS_KEY("Backup");
    const QString KEY_FILELOGGER_DELETEOLD = FILELOGGER_SETTINGS_KEY("DeleteOld");
    const QString KEY_FILELOGGER_MAXSIZE = FILELOGGER_SETTINGS_KEY("MaxSize");
    const QString KEY_FILELOGGER_AGE = FILELOGGER_SETTINGS_KEY("Age");
    const QString KEY_FILELOGGER_AGETYPE = FILELOGGER_SETTINGS_KEY("AgeType");
    
    // Other class properties keys (just an example for this snippet)
#define OTHER_CLASS_SETTINGS_KEY(name) SETTINGS_KEY("OtherClass/") name
    const QString KEY_OTHER_CLASS_ENABLED = OTHER_CLASS_SETTINGS_KEY("Enabled");

    //just a shortcut
    inline SettingsStorage *settings() { return  SettingsStorage::instance(); }

    const QString LOG_FOLDER("logs");
    const char PARAMS_SEPARATOR[] = "|";
}
```

glassez commented:
>Would the following be considered ok or does it constitute preprocessor abuse?

I don't think it will tire the preprocessor

sledgehammer999 commented:
Thanks for reviewing. Final suggestions implemented. Pushed to master.
